packages/react-client/src/ReactFlightClient.js

```diff
@@ -907,26 +907,41 @@ function waitForReference<T>(
       }
       value = value[path[i]];
     }
-    parentObject[key] = map(response, value);
+    const mappedValue = map(response, value);
+    parentObject[key] = mappedValue;
 
     // If this is the root object for a model reference, where `handler.value`
     // is a stale `null`, the resolved value can be used directly.
     if (key === '' && handler.value === null) {
-      handler.value = parentObject[key];
+      handler.value = mappedValue;
     }
 
-    // If the parent object is an unparsed React element tuple and its outlined
-    // props have now been resolved, we also need to update the props of the
-    // parsed element object (i.e. handler.value).
+    // If the parent object is an unparsed React element tuple, we also need to
+    // update the props, owner, and stack of the parsed element object (i.e.
+    // handler.value).
     if (
       parentObject[0] === REACT_ELEMENT_TYPE &&
-      key === '3' &&
       typeof handler.value === 'object' &&
       handler.value !== null &&
-      handler.value.$$typeof === REACT_ELEMENT_TYPE &&
-      handler.value.props === null
+      handler.value.$$typeof === REACT_ELEMENT_TYPE
     ) {
-      handler.value.props = parentObject[key];
+      const element: any = handler.value;
+      switch (key) {
```

Comment:
What about "key" and "type"? E.g. when they're blocked client references.

Reply:
In my testing, those were just normal client references, using your example but adapted for `clientExports`:

```
"use client";

export const key = "hi";
export const type = "div";
```

```
import { key, type as Component } from …

<Component key={key} />
```

Reply:
But what if they're not yet resolved? What happens then?

Reply:
That's a lazy ID that's being resolved, not a reference ID. In other words, the `default` case of `parseModelString` is not used, and therefore `getOutlinedModel` is not called which would call `waitForReference`. So this code is not used at all. But maybe you're saying we also need to do this dance elsewhere? 🤔

Reply:
actually, we get this error: `Error: Element type is invalid. Received a promise that resolves to: div. Lazy element type must resolve to a class or function.`

and the `key` is the stringified client reference proxy function, it seems: `"function() {throw new Error(\"Attempted to call key() from the server but key is on the client. It's not possible to invoke a client function from the server, it can only be rendered as a Component or passed to props of a Client Component.\");}"`